packages/remirror__extension-link/src/link-extension.ts

```diff
@@ -575,59 +609,113 @@ export class LinkExtension extends MarkExtension<LinkOptions> {
 
         const { updateLink, removeLink } = this.store.chain(tr);
 
-        changes.forEach(({ from, to, prevFrom, prevTo }) => {
-          // Remove auto links that are no longer valid
-          this.getLinkMarksInRange(prevState.doc, prevFrom, prevTo, true).forEach((prevMark) => {
-            const newFrom = mapping.map(prevMark.from);
-            const newTo = mapping.map(prevMark.to);
-            const newMarks = this.getLinkMarksInRange(doc, newFrom, newTo, true);
-
-            newMarks.forEach((newMark) => {
-              const wasLink = this._autoLinkRegexNonGlobal?.test(prevMark.text);
-              const isLink = this._autoLinkRegexNonGlobal?.test(newMark.text);
-
-              if (wasLink && !isLink) {
-                removeLink({ from: newMark.from, to: newMark.to }).tr();
-              }
-            });
-          });
-
+        changes.forEach(({ prevFrom, prevTo, from, to }) => {
           // Store all the callbacks we need to make
           const onUpdateCallbacks: Array<Pick<EventMeta, 'range' | 'attrs'> & { text: string }> =
             [];
 
-          // Find text that can be auto linked
-          tr.doc.nodesBetween(from, to, (node, pos) => {
-            if (!node.isTextblock || !node.type.allowsMarkType(this.type)) {
-              return;
-            }
+          // Check if node was split into two by `Enter` key press
+          const isNodeSeparated = to - from === 2;
+
+          // Get previous links
+          const prevMarks = this.getLinkMarksInRange(prevState.doc, prevFrom, prevTo, true)
+            .filter((item) => item.mark.type === this.type)
+            .map(({ from, to, text }) => ({
+              mappedFrom: mapping.map(from),
+              mappedTo: mapping.map(to),
+              text,
+              from,
+              to,
+            }));
+
+          // Check if links need to be removed or updated.
+          prevMarks.forEach(
+            ({ mappedFrom: newFrom, mappedTo: newTo, from: prevMarkFrom, to: prevMarkTo }, i) =>
+              this.getLinkMarksInRange(doc, newFrom, newTo, true)
+                .filter((item) => item.mark.type === this.type)
+                .forEach((newMark) => {
+                  const prevLinkText = prevState.doc.textBetween(
```

Comment:
There is a [`textBetween` util](https://github.com/remirror/remirror/blob/ac2019fc59b8caf1f1f8eaaa3163b5a15a459c5c/packages/remirror__core-utils/src/core-utils.ts#L1472-L1502) that may be of use here.

```ts
import { textBetween } from '@remirror/core';

const prevLinkText = textBetween({
  doc: prevState.doc,
  from: prevMarkFrom,
  to: prevMarkTo,
});
```

Reply:
I think this util returns an array and special characters are escaped. Do you mind if we keep it as is?